### PR TITLE
Add CSV download for a single metric

### PR DESCRIPTION
**Pull request by KyleMacPherson**
EDIT: DO NOT MERGE IN CURRENT STATE, WILL BREAK SOME GA METRICS

# What
Add CSV download for a single metric. This involves adding a new metric controller and renaming our existing metric controller to single page controller. This was necessary because using the splat operator to assign the base path param in the old metrics (now single page) controller did not allow us to extend the controller.

Following the pattern established by the csv download work for the index page I have created a new single page csv presenter and made use of some of the existing `export_to_csv` logic.

# Why
In order to let publishers or editors be able to inspect individual metric data for a page in a spreadsheet.

# Screenshots
screen shot 2018-12-11 at 16 26 13

---
# Review Checklist
* [ ] Changes in scope.
* [ ] Added/updated unit tests.
* [ ] Added/updated feature tests.
* [ ] Added/updated relevant documentation.
* [x] Added to Trello card.

### CODE REVIEW

File: app/presenters/metrics_csv_presenter.rb [hunk 1]
```diff
@@ -0,0 +1,28 @@
+require 'csv'
+
+class MetricsCSVPresenter
```
Comment (pmanrubia) [inline app/presenters/metrics_csv_presenter.rb]: Can you add test for this class?

File: spec/support/content_data_api.rb [hunk 2]
```diff
@@ -119,6 +119,12 @@ def content_data_api_has_document_types
         stub_request(:get, url).to_return(status: 200, body: body)
       end
 
+      def content_data_api_has_csv
```
Comment (pmanrubia) [inline spec/support/content_data_api.rb]: We should be explicit when setting the fixture for a test. Could you please extract the body?

File: spec/features/csv_download_spec.rb [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    GDS::SSO.test_user = build(:user)
+    stub_metrics_page(base_path: 'base/path', time_period: :past_30_days)
+    visit '/single_page/base/path'
+    content_data_api_has_csv
+  end
+
+  it 'downloads csv' do
+    visit '/single_page/base/path?date_range=past-30-days'
+    click_link 'Download Unique pageviews CSV'
+    expect(CSV.parse(page.body)).to eq [%w(Date Value), [date2, "1000"], [date3, "2000"], [date1, "3000"]]
```
Comment (pmanrubia) [inline spec/features/csv_download_spec.rb]: It is not clear where do 1000, 2000 and 3000 comes from. What about being explicit when setting up the fixture?

Comment (KyleMacPherson) [reply spec/features/csv_download_spec.rb]: 👍

File: spec/support/content_data_api.rb (hunk 2 above)
Comment (KyleMacPherson) [reply spec/support/content_data_api.rb]: 👍

File: app/presenters/metrics_csv_presenter.rb (hunk 1 above)
Comment (KyleMacPherson) [reply app/presenters/metrics_csv_presenter.rb]: 👍

**Comment by theseanything** [comment]
[(5) Add more fields in index page CSV download](https://trello.com/c/lyOMgJds/898-5-add-more-fields-in-index-page-csv-download)

File: spec/support/content_data_api.rb [hunk 4]
```diff
@@ -119,12 +119,6 @@ def content_data_api_has_document_types
         stub_request(:get, url).to_return(status: 200, body: body)
       end
 
-      def content_data_api_has_csv
```
Comment (pmanrubia) [inline spec/support/content_data_api.rb]: We should be explicit when setting up the tests. It is better not to hide the response with the method, so it can be reused in different places:

I would suggest something like:

```ruby
  def content_data_api_has_csv(columns: [], values: [])
  end
```

And then use it with:

```ruby
content_data_api_has_csv(
  columns: %w(Date Value), 
  values: [['10-10-2018',1000],[['11-10-2018',1100]])
```
To be consistent with what we already have, I would put this method in `content_data_api`

Comment (KyleMacPherson) [reply spec/support/content_data_api.rb]: Good idea! 👍

File: app/views/content/index.html.erb [hunk 5]
```diff
@@ -20,28 +20,33 @@
...
               {
                 value: "past-year",
-                text: t("metrics.show.time_periods.past-year.leading"),
+                text: t("single_page.show.time_periods.past-year.leading"),
                 hint_text: "#{(1.year.ago).strftime("%-d %B %Y")} to #{Date.yesterday.strftime("%-d %B %Y")}"
+              },
+              {
+                value: "past-2-years",
+                text: t("single_page.show.time_periods.past-2-years.leading"),
+                hint_text: "#{(2.years.ago).strftime("%-d %B %Y")} to #{Date.yesterday.strftime("%-d %B %Y")}"
```
Comment (cbaines) [inline app/views/content/index.html.erb]: Is this change intentional? I remember something about this time range being removed, but I haven't checked.

Comment (KyleMacPherson) [reply app/views/content/index.html.erb]: Ah well spotted, thanks Chris.

File: app/controllers/metrics_controller.rb [hunk 6]
```diff
@@ -1,22 +1,18 @@
+require 'csv'
```
Comment (cbaines) [inline app/controllers/metrics_controller.rb]: This looks redundant, as `CSV` isn't used in this file.

File: app/presenters/single_content_item_presenter.rb [hunk 7]
```diff
@@ -127,6 +127,12 @@ def link_text(metric_name)
     I18n.t("metrics.#{metric_name}.title").try(:downcase)
   end
 
+  def csv_url(metric_name)
+    from = @date_range.from
+    to = @date_range.to
+    "/metrics#{base_path}?metric_name=#{metric_name}&from=#{from}&to=#{to}.csv"
```
Comment (cbaines) [inline app/presenters/single_content_item_presenter.rb]: Perhaps just remove the `.csv` bit from the end here? I don't think it adds much.

File: lib/gds_api/content_data_api.rb [hunk 8]
```diff
@@ -64,7 +64,7 @@ def content_items_url(date_range, organisation_id, document_type, page, page_siz
   end
 
   def single_page_url(base_path, from, to)
-    "#{content_data_api_endpoint}/single_page/#{base_path}#{query_string(from: from, to: to)}"
+    "#{content_data_api_endpoint}/metrics/#{base_path}#{query_string(from: from, to: to)}"
```
Comment (cbaines) [inline lib/gds_api/content_data_api.rb]: Is this change right? It seems odd to me.

**Comment by pmanrubia** [comment]
I am closing this PR. It is difficult to follow as the commits are not clear and don't seem to be green. Also Kyle is not in the team any more to ask for clarifications.
